### PR TITLE
feat: support previous names

### PR DESCRIPTION
**Pull request by akaJSaunders**
### JIRA

[Jira Ticket](https://companieshouse.atlassian.net/browse/SR-83)

### Work Checklist

- [ ] Tests added where applicable
- [ ] Test coverage of new code meets target of 80%
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

---

### Merge Instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/ApplicantDetailsController.ts [hunk 1]
```diff
@@ -22,6 +24,11 @@ export class ApplicantDetailsController {
 
   public processForm = async (req: Request, res: Response, next: NextFunction) => {
     const validationResult: ValidationResult = await this.validator.validate(req);
+
+    if (req.body.hasPreviousName === YesNo.no){
+      req.body.previousName = '';
```
Comment (rhysbarrett) [inline src/controllers/ApplicantDetailsController.ts:29]: The validation schema appears to use conditional validation on this field, and the field is set as optional in the model. Does it need setting as an empty string here?

Comment (akaJSaunders) [reply src/controllers/ApplicantDetailsController.ts:29]: So imagine a user initially put a previous name, and then proceeded through the journey, but then returned back to that page and selected no, we want to make sure we clear the information they might have put in previousName.

File: src/validators/schema/ApplicantDetailsSchema.ts [hunk 2]
```diff
@@ -14,6 +17,24 @@ export const schema = Joi.object({
...
+    .valid(YesNo.yes, YesNo.no)
+    .messages({
+      'any.required': hasPreviousNameMissingMessage,
+      'any.only': hasPreviousNameMissingMessage,
+      'string.base': hasPreviousNameMissingMessage,
+      'string.empty': hasPreviousNameMissingMessage
+    }),
+  previousName: Joi.when('hasPreviousName', {
+      is: Joi.valid(YesNo.yes).exist(),
+      then: Joi.string().required().pattern(/\w+/).messages({
```
Comment (rhysbarrett) [inline src/validators/schema/ApplicantDetailsSchema.ts:31]: Feel free to use the `basicString` schema item, as seen in some of the other schemas. For example: `then: basicString(previousNameMissingMessage)`.

File: test/controllers/ApplicantDetailsController.test.ts [hunk 3]
```diff
@@ -76,7 +113,10 @@ describe('ApplicantDetailsController', () => {
 
       await request(app).post(APPLICANT_DETAILS_PAGE_URI).send({
         fullName: 'John Doe',
+        hasPreviousName: 'yes',
+        previousName: 'test_name',
         emailAddress: 'test.com'
+
```
Comment (rhysbarrett) [inline test/controllers/ApplicantDetailsController.test.ts:119]: whitespace

**Review by rhysbarrett — APPROVED** [review]

File: src/controllers/ApplicantDetailsController.ts (hunk 1 above)
Comment (rhysbarrett) [reply src/controllers/ApplicantDetailsController.ts:29]: Ah yes, I see. I suppose the only difference between how I did it in Appeals and how you're doing it here is that I acted upon the data object after validation, while you're acting on the form data prior to validation. Nice!

File: src/validators/schema/ApplicantDetailsSchema.ts (hunk 2 above)
Comment (akaJSaunders) [reply src/validators/schema/ApplicantDetailsSchema.ts]: awesome! hadnt noticed we had that

**Review by rhysbarrett — APPROVED** [review]

**Pull request merged**